pycbc/io/hdf.py

```diff
@@ -56,52 +67,89 @@ def select(self, fcn, *args, **kwds):
             element is an array of indices of elements passing the function.
 
         >>> f = HFile(filename)
-        >>> snr = f.select(lambda snr: snr > 6, 'H1/snr')
+        >>> snr = f.select(lambda snr: snr > 6, 'snr', group='H1')
         """
 
+        grp = kwds.get('group', None)
         # get references to each array
+        data_grp = self[grp]
         refs = {}
         data = {}
+
+        # Required datasets are either the arguments requested, we also add
+        # required datasets if the value requested is a known ranking
+        # (or a combination)
+        req_dsets = []
+        for a in args:
+            if a in data_grp:
+                req_dsets.append(a)
+            elif a in ranking.required_datasets:
+                # Get the required datasets from the defined list
+                req_dsets += ranking.required_datasets[a]
+            else:
+                raise ValueError(f"Cannot work out required datasets for {a}")
+
+        # remove any duplicates from req_dsets
+        req_dsets = list(set(req_dsets))
+
+        # Get the pointers to the h5py Datasets:
+        for ds in req_dsets:
+            refs[ds] = data_grp[ds]
+
+        # This will be the outputs:
         for arg in args:
-            refs[arg] = self[arg]
             data[arg] = []
 
         return_indices = kwds.get('return_indices', False)
+        indices_only = kwds.get('indices_only', False)
         indices = np.array([], dtype=np.uint64)
 
         # To conserve memory read the array in chunks
         chunksize = kwds.get('chunksize', int(1e6))
-        size = len(refs[arg])
+        size = len(refs[ds])
+
+        mask = kwds.get('premask', np.ones(size, dtype=bool))
+        if not mask.dtype == bool:
+            # mask is an array of indices rather than booleans,
+            # make it a bool array
+            new_mask = np.zeros(size, dtype=bool)
+            new_mask[mask] = True
+            mask = new_mask
 
         i = 0
         while i < size:
             r = i + chunksize if i + chunksize < size else size
 
-            # Read each chunk's worth of data and find where it passes the function
-            partial = [refs[arg][i:r] for arg in args]
+            # Read each chunk's worth of data and find where it passes
+            # the function
+            partial_data = {arg: refs[arg][i:r][mask[i:r]]
+                            for arg in req_dsets}
+            partial = [partial_data[a] if a in partial_data
+                       else ranking.get_sngls_ranking_from_trigs(
```

Comment:
@GarethCabournDavies This seems like a odd shoehorn into this class to me. The idea is that this class should be general and only really build upon h5py / numpy and so can be used for general purposes where more careful access of file informatoin is needed.  Is there a way to avoid explicitly calling a module from ranking.py? Perhaps by extending the options to this function?

Reply:
@GarethCabournDavies Ok, I think one approach would be to add something like a 'derived' keyword where on can provide a dictionary -> function mapping (could b lambda function to allow more complicated function interface like you might have with the ranking one. 

So something like the following input. 

derived = {'derived_name': function} 

And in this case, the 'required' keys should just be expanded with the ones you know your functions need. I see you added this data to ranking.py, so I think where you are calling the select method from should have this information.

Reply:
That makes sense - I think this will still require the dictionary setup I used for passing to the derived (ranking) functions, as otherwise there could be duplicated datasets which we should avoid, but this does add a layer of generality

Reply:
The select() function has been updated to be agnistic of any other module.

SingleDetTriggers still calls through for defined ranking statistics - as SingleDetTriggers already has this happening for certain attributes, I thought this should be okay still

Reply:
SingleDetTriggers is specialized so I think it makes sense there.